Add support to dot notation in toHaveKey expectation

fabio-ivona opened the pull request:
This PR wants to add support for dot notation when giving the array key to test with `expect()->toHaveKey()` expectation

this will allow to write a test like this one:

```php
it('has nested key', function () {
    $array = [
        'a' => 1,
        'b' => 2,
        'c' => [
            'd' => 3,
        ],
        'e.f' => 4,
    ];
    expect($array)->toHaveKey('c.d');
    expect($array)->toHaveKey('c.d', 3);
    expect($array)->toHaveKeys(['c.d', 'a']);
    expect($array)->toHaveKey('e.f');
});
```

As you can see in the last expectation of the example, In order to not be a breaking change, this implementation still allows to address an array key containing a dot in the string.

**some credits:** 

- to parse dot notated keys I added an helper `Arr` static class,  taking inspiration for `has()` and `get()` methods implementation from the [laravel/framework](https://github.com/laravel/framework) project

- @olivernybroe gave me the nice idea of allowing arrays with dots in it to obtain a non-breaking change

**bonus:** while adding these new tests cases to the `tests/Expect/toHaveKey.php` tests, I noticed that the second assertion of `toHaveKey($key, $value)` expectation (assert that the array has that specific value for the given key) was never tested, so added tests for it as well

fabio-ivona committed:
enables dot notation in toHaveKey expectation

fabio-ivona committed:
enables dot notation in toHaveKey expectation

fabio-ivona committed:
adds tests for dot notation in ->toHaveKey expectation

fabio-ivona committed:
adds tests for dot notation in ->toHaveKeys expectation

fabio-ivona committed:
Merge remote-tracking branch 'origin/dot-notation-in-toHaveKey' into dot-notation-in-toHaveKey

# Conflicts:
#	src/Helpers/Arr.php

fabio-ivona committed:
remove only() from test

olivernybroe commented on src/Expectation.php:
```diff
@@ -522,10 +523,10 @@ public function toHaveKey($key, $value = null): Expectation
         }
 
-        Assert::assertArrayHasKey($key, $array);
+        Assert::assertTrue(Arr::has($array, $key));
```
As this assertions is now an `assertTrue`, the error message will not be as clear, maybe we should add a custom message?

fabio-ivona replied:
I've resolved adding a custom message:

`Assert::assertTrue(Arr::has($array, $key), "Failed asserting that an array has the key '$key'");`

but it will generate this:

> Failed asserting that an array has the key 'foo'
> Failed asserting that false is true.

Any idea on how to avoid the second part of the message? I think it is quite useless, but it is added from `Assert::assertTrue`

fabio-ivona replied:
I've implemented a possible workaround wrapping `Assert::assertTrue(Arr::has($array, $key));` in a try-catch block and generating a custom `ExpectationFailedException`:

```php
try {
    Assert::assertTrue(Arr::has($array, $key));
} catch (ExpectationFailedException $exception) {
    throw new ExpectationFailedException(
        "Failed asserting that an array has the key '$key'",
        $exception->getComparisonFailure()
    );
}
```

olivernybroe commented on src/Helpers/Arr.php:
```diff
@@ -0,0 +1,68 @@
+namespace Pest\Expectations\Helpers;
+
+/**
+ * Class Arr.
```
Let's remove the `Class Arr` comment, it's not really helpful

olivernybroe commented on src/Helpers/Arr.php:
```diff
@@ -0,0 +1,68 @@
+ * belongs to the Arr helper of laravel/framework project
+ * (https://github.com/laravel/framework)
+ */
+final class Arr
```
Add `internal` annotation to this class as we don't to support this class as kind of public API for pest 👍

olivernybroe reviewed (approved):
Looks amazing and simple, just had some minor comments 👍

fabio-ivona committed:
fixes Arr.php doc comments

fabio-ivona committed:
adds a clearer failure message for toHaveKey expectation

fabio-ivona committed:
cleans up toHaveKey expectation failure message

fabio-ivona commented:
Just updated code to address @olivernybroe comments

olivernybroe reviewed (commented):
Really nice, could we add a test for the error message also, in case someone tries to refactor it to `assertTrue` with a custom message, but doesn't realise that it adds two lines 👍

fabio-ivona committed:
added tests for toHaveKey failure message and refactored other toHaveKey tests

olivernybroe reviewed (approved):
LGTM! 

Thanks for the great work ❤️

fabio-ivona commented:
Done, I updated _failures_ tests adding the exceptionMessage check:

`->throws(ExpectationFailedException::class, "Failed asserting that an array has the key 'foo'")`

also I refactored the whole test file, using a dataset would not have been compatible with the `->throws(...)` directive, as the message would change for each dataset

fabio-ivona commented:
hmm, seems there is an issue with composer dependencies.. is it something I can solve myself?

olivernybroe commented:
hmm @owenvoke could you help out with the composer issues that @fabio-ivona is having? 😃

owenvoke commented:
Oh, it's the old circular dependencies issue. 😦 I'm not actually sure how to resolve that. I think I opened PRs a while back that helped fix it by basically adding `COMPOSER_ROOT_VERSION: 1.x` to the workflow environment, but not sure where that went.

I'll open a PR in a bit to see if I can get it fixed.

lukeraymonddowning commented:
@owenvoke I've fixed it in #16

nunomaduro commented:
This tests issue will be fixed once we merge all plugins on the core. @owenvoke Is taking care of that.

fabio-ivona commented:
That's fine for me

At this point I don't think this PR needs more work from me, let me know if you need something else ;)

nunomaduro commented:
@lukeraymonddowning Can you please perform this pull request on the core?
